ArcGISRuntimeSDKQt_CppSamples/Analysis/ViewshedGeoElement/ViewshedGeoElement.cpp

```diff
@@ -0,0 +1,184 @@
+// [WriteFile Name=ViewshedGeoElement, Category=Analysis]
+// [Legal]
+// Copyright 2017 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#include "ViewshedGeoElement.h"
+
+#include "ArcGISTiledElevationSource.h"
+#include "Scene.h"
+#include "SceneQuickView.h"
+#include "Camera.h"
+#include "Point.h"
+#include "Viewpoint.h"
+#include "ArcGISSceneLayer.h"
+#include "SimpleRenderer.h"
+#include "ModelSceneSymbol.h"
+#include "GeoElementViewshed.h"
+#include "GeometryEngine.h"
+#include "OrbitGeoElementCameraController.h"
+
+#include <QQmlProperty>
+#include <QTimer>
+#include <QString>
+#include <QUrl>
+#include <QVariant>
+#include <QList>
+
+using namespace Esri::ArcGISRuntime;
+
+ViewshedGeoElement::ViewshedGeoElement(QQuickItem* parent /* = nullptr */):
+  QQuickItem(parent)
+{
+}
+
+void ViewshedGeoElement::init()
+{
+  // Register classes for QML
+  qmlRegisterType<SceneQuickView>("Esri.Samples", 1, 0, "SceneView");
+  qmlRegisterType<ViewshedGeoElement>("Esri.Samples", 1, 0, "ViewshedGeoElementSample");
+}
+
+void ViewshedGeoElement::componentComplete()
+{
+  QQuickItem::componentComplete();
+
+  // Create a scene and give it to the SceneView
+  m_sceneView = findChild<SceneQuickView*>("sceneView");
+  Scene* scene = new Scene(Basemap::imagery(this), this);
+
+  // Create a surface
+  Surface* surface = new Surface(this);
+  surface->elevationSources()->append(
+        new ArcGISTiledElevationSource(
+          QUrl("http://scene.arcgis.com/arcgis/rest/services/BREST_DTM_1M/ImageServer"),
+          this));
+  scene->setBaseSurface(surface);
+
+  // Add a SceneLayer
+  ArcGISSceneLayer* sceneLayer = new ArcGISSceneLayer(QUrl("http://tiles.arcgis.com/tiles/P3ePLMYs2RVChkJx/arcgis/rest/services/Buildings_Brest/SceneServer/layers/0"), this);
+  scene->operationalLayers()->append(sceneLayer);
+
+  // Add an AnalysisOverlay
+  m_analysisOverlay = new AnalysisOverlay(this);
+  m_sceneView->analysisOverlays()->append(m_analysisOverlay);
+
+  // Add a GraphicsOverlay
+  createGraphicsOverlay();
+
+  // Create a Graphic
+  createGraphic();
+
+  // Create the GeoElementViewshed
+  const double horizontalAngle = 90.0;
+  const double verticalAngle = 25.0;
+  const double minDistance = 1.0;
+  const double maxDistance = 250.0;
+  const double headingOffset = 0.0;
+  const double pitchOffset = 0.0;
```

Comment:
It would make sense to me to add values for `offsetX`, `offsetY`, and `offsetZ`.  But not sure if that's in the design.